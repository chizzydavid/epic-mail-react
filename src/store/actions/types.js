export const INIT_AUTH_REQUEST= 'INIT_AUTH_REQUEST';

export const SET_CURRENT_USER = 'SET_CURRENT_USER';

export const AUTH_ERROR = 'AUTH_ERROR';

export const BEGIN_MSG_REQUEST= 'BEGIN_MSG_REQUEST';

export const END_MSG_REQUEST= 'END_MSG_REQUEST';

export const SET_RECEIVED_MESSAGES = 'SET_RECEIVED_MESSAGES';

export const SET_UNREAD_MESSAGES = 'SET_UNREAD_MESSAGES';

export const SET_READ_MESSAGES = 'SET_READ_MESSAGES';

export const SET_SENT_MESSAGES = 'SET_SENT_MESSAGES';

export const SET_DRAFTS = 'SET_DRAFTS';

export const MSG_FEEDBACK = 'MSG_FEEDBACK';

export const MSG_ERROR = 'MSG_ERROR';

export const SET_ACTIVE_TAB = 'SET_ACTIVE_TAB';

export const INIT_GET_REQUEST = 'INIT_GET_REQUEST';
