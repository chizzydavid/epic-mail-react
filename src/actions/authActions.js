import { 
  SET_CURRENT_USER, 
  AUTH_ERROR, 
  INIT_AUTH_REQUEST 
} from './types';
import axios from 'axios';
import setAuthToken from '../utils/setAuthToken';

const authUrl = `https://chizzy-epicmail.herokuapp.com/api/v2/auth/`;

export const signUpUser = (payload) => dispatch => {
  dispatch({type: INIT_AUTH_REQUEST}); 

  axios.post(`${authUrl}signup`, payload)
    .then((res) => {
      const { token, user } = res.data.data;
      localStorage.setItem('token', token);

      setAuthToken(token);
      dispatch(setCurrentUser(user));
    })
    .catch(err => {
      console.log(err);
      dispatch({
        type: AUTH_ERROR,
        payload: err.response.data
      });
    });
}

export const setCurrentUser = payload => {
  return {
    type: SET_CURRENT_USER,
    payload: payload
  }
}
