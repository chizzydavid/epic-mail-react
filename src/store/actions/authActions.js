import { 
  SET_CURRENT_USER, 
  AUTH_ERROR, 
  INIT_AUTH_REQUEST 
} from './types';
import axios from 'axios';
import setAuthToken from '../../utils/setAuthToken';

const authUrl = `https://chizzy-epicmail.herokuapp.com/api/v2/auth/`;

export const signUpUser = (payload) => dispatch => {
  dispatch({type: INIT_AUTH_REQUEST}); 

  return axios.post(`${authUrl}signup`, payload)
    .then((res) => {
      const { token, user } = res.data.data;
      localStorage.setItem('token', token);

      setAuthToken(token);
      dispatch(setCurrentUser(user));
    })
    .catch(err => {
      dispatch({
        type: AUTH_ERROR,
        payload: err.response.data
      });
    });
}

export const setCurrentUser = payload => {
  return {
    type: SET_CURRENT_USER,
    payload: payload
  }
}

export const loginUser = (payload) => dispatch => {
  dispatch({type: INIT_AUTH_REQUEST});

  return axios.post(`${authUrl}login`, payload)
    .then((res) => {
      const { token, user } = res.data.data;
      localStorage.setItem('token', token);

      setAuthToken(token);
      dispatch(setCurrentUser(user));
    })
    .catch(err => {
      dispatch({
        type: AUTH_ERROR,
        payload: err.response.data
      })
    }) 
}

