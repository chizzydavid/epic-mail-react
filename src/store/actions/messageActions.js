import {
  BEGIN_MSG_REQUEST,
  MSG_ERROR,
  MSG_FEEDBACK,
  SEND_MSG_FEEDBACK,
  SET_RECEIVED_MESSAGES,
  SET_UNREAD_MESSAGES,
  SET_READ_MESSAGES,
  SET_SENT_MESSAGES,
  SET_DRAFTS,
  SET_ACTIVE_TAB,
  REMOVE_MESSAGE_ALL
} from './types';
import { BASE_URL } from '../constants';
import axios from 'axios';

export const getMessages = (category) => dispatch => {
  dispatch({type: BEGIN_MSG_REQUEST}); 
  const msgCategory = category === 'all' ? '' : category;

  return axios.get(`${BASE_URL}messages/${msgCategory}`)
    .then(res => {
      dispatch(setActiveMsgTab(category));
      if (res.data.message) {
        dispatch(messageFeedback(res.data.message));
        return;
      }
      dispatch(setMessages(category, res.data.data));
    })
    .catch(err => {
      dispatch(messageError(err.response.data));
    });
}

export const sendMessage = (payload) => dispatch => {
  dispatch({type: BEGIN_MSG_REQUEST}); 

  axios.post(`${BASE_URL}messages/`, payload)
    .then((res) => {
      if (res.data.status === 201) {
        dispatch(sendMessageFeedback('Message sent successfully.'))
      }
      dispatch(setMessages('sent', res.data));
    })
    .catch(err => {
      dispatch(messageError(err.response.data))
    });
}

export const deleteMessage = (messageId) => dispatch => {
  dispatch({type: BEGIN_MSG_REQUEST}); 

  axios.delete(`${BASE_URL}messages/${messageId}`)
    .then((res) => {
      if (res.data.status === 200) {
        dispatch({
          type: REMOVE_MESSAGE_ALL,
          payload: messageId,
        });
      }
    })
    .catch(err => {
      dispatch(messageError(err.response.data))
    });
}

export const setActiveMsgTab = category  => {
  return {
    type: SET_ACTIVE_TAB,
    payload: category
  }
}

export const setMessages = (category, payload) => {
  let type = category === 'all' ? SET_RECEIVED_MESSAGES : 
    category === 'unread' ? SET_UNREAD_MESSAGES : 
    category === 'read' ? SET_READ_MESSAGES : 
    category === 'sent' ? SET_SENT_MESSAGES : 
    category === 'drafts' ? SET_DRAFTS : '';

  return {
    type: type,
    payload: payload
  }
}

export const messageError = payload => {
  return {
    type: MSG_ERROR,
    payload: payload
  }
}

export const messageFeedback = payload => {
  return {
    type: MSG_FEEDBACK,
    payload: payload
  }
}

export const sendMessageFeedback = payload => {
  return {
    type: SEND_MSG_FEEDBACK,
    payload: payload
  }
}
