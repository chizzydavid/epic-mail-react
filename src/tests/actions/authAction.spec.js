/* eslint-disable arrow-parens */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nock from 'nock';
import { signUpUser, loginUser } from '../../store/actions/authActions';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const mockUser = {
  firstName: 'Chizindu',
  lastName: 'David',
  userName: 'gr',
  email: 'chizzydavid@gmail.com',
  password: 'password',
  confirmPassword: 'password',
};

const userData = {
  email: 'chizzydavid@gmail.com',
  password: 'password',
};


describe('authAction', () => {
  let store;

  beforeEach(() => {
    store = mockStore({});
  });
  afterEach(() => {
    // clear all HTTP mocks after each test
    nock.cleanAll();
  });

  it('should decode user token when registerUser is successful', done => {
    nock('https://localhost:3000')
      .post('/api/users')
      .reply(201, {
        message: 'user registration was successful',
        user: {
          id: 24,
          firstName: 'chizzy',
          lastName: 'david',
          email: 'chizzydavid@gmail.com',
          userName: 'chizzy',
        },
        token:
          'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOjI0LCJlbWFpbCI6InByaW5jZXNzQGdtYWlsLmNvbSIsInVzZXJOYW1lIjoiZ3IiLCJpYXQiOjE1NjI1Mjg2NTcsImV4cCI6MTU2MzEzMzQ1N30.oWthtPvSh-zz4RwgHZsJtdxpjhHlUKix0oK1I9nqkOA',
      });
    return store.dispatch(signUpUser(mockUser)).then(() => {
      expect(store.getActions()).toMatchSnapshot();
      done();
    });
  });

  it('should handle errors returned from the request', done => {
    const errorMessage = 'This username is already in use';
    nock('https://localhost:3000')
      .post('/api/users')
      .reply(400, { error: errorMessage });

    return store.dispatch(signUpUser({})).then(() => {
      expect(store.getActions()).toMatchSnapshot();
      done();
    });
  });


  it('Logs in a user', () => {
    nock('https://localhost:3000')
      .post('/api/users', userData)
      .reply(200, {
        message: 'user registration was successful',
        user: {
          id: 24,
          firstName: 'chizzy',
          lastName: 'david',
          email: 'chizzydavid@gmail.com',
          userName: 'chizzy',
        },
        token:
          'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOjI0LCJlbWFpbCI6InByaW5jZXNzQGdtYWlsLmNvbSIsInVzZXJOYW1lIjoiZ3IiLCJpYXQiOjE1NjI1Mjg2NTcsImV4cCI6MTU2MzEzMzQ1N30.oWthtPvSh-zz4RwgHZsJtdxpjhHlUKix0oK1I9nqkOA',
      });

    return store.dispatch(loginUser(userData))
      .then(() => {
        const expectedActions = ['INIT_AUTH_REQUEST', 'SET_CURRENT_USER'];
        const dispatchedAction = store.getActions();
        const actionTypes = dispatchedAction.map(action => action.type);
        expect(actionTypes).toEqual(expectedActions);
        // expect(store.getActions()).toMatchSnapshot();

      });
  });

  it('Logs in a user', () => {
    nock('https://localhost:3000')
      .post('/api/users', {})
      .replyWithError();

    return store.dispatch(loginUser({}))
      .then(() => {
        const expectedActions = ['INIT_AUTH_REQUEST', 'AUTH_ERROR'];
        const dispatchedAction = store.getActions();

        const actionTypes = dispatchedAction.map(action => action.type);
        expect(actionTypes).toEqual(expectedActions);
        // expect(store.getActions()).toMatchSnapshot();
      });
  });
});
