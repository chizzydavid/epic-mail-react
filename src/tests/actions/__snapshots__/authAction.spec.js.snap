// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authAction should decode user token when registerUser is successful 1`] = `
Array [
  Object {
    "type": "INIT_AUTH_REQUEST",
  },
  Object {
    "payload": Object {
      "error": "Email already exists.",
      "status": 400,
    },
    "type": "AUTH_ERROR",
  },
]
`;

exports[`authAction should handle errors returned from the request 1`] = `
Array [
  Object {
    "type": "INIT_AUTH_REQUEST",
  },
  Object {
    "payload": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Internal Server Error</pre>
</body>
</html>
",
    "type": "AUTH_ERROR",
  },
]
`;
